{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70023de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option(\"display.max_rows\", 10)\n",
    "pd.set_option(\"display.max_columns\", 10)\n",
    "pd.set_option(\"display.max_colwidth\", 50)\n",
    "pd.set_option(\"display.width\", 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b2c1d76d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# local paths\n",
    "data_dir_raw = 'data/raw'\n",
    "data_dir_proc = 'data/processed'\n",
    "Path(data_dir_proc).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e1fabb9",
   "metadata": {},
   "source": [
    "# load raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "451c0903",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           TCGA-69-7978-01  TCGA-62-8399-01  TCGA-78-7539-01  TCGA-50-5931-11  TCGA-73-4658-01  \\\n",
      "sample                                                                                           \n",
      "ARHGEF10L           9.9898          10.4257           9.6264           8.6835           9.2078   \n",
      "HIF3A               4.2598          11.6239           9.1362           9.4824           5.0288   \n",
      "RNF17               0.4181           0.0000           1.1231           0.8221           0.0000   \n",
      "RNF10              10.3657          11.5489          11.6692          11.7341          11.6209   \n",
      "RNF11              11.1718          11.0200          10.4679          11.6787          11.3414   \n",
      "\n",
      "           ...  TCGA-55-7727-01  TCGA-91-6831-01  TCGA-MN-A4N4-01  TCGA-55-8302-01  \\\n",
      "sample     ...                                                                       \n",
      "ARHGEF10L  ...           7.2428           8.8388           9.9341          10.1696   \n",
      "HIF3A      ...           7.5416           3.5613           8.3457           5.5364   \n",
      "RNF17      ...           0.9392           0.0000           0.9715           0.0000   \n",
      "RNF10      ...          11.8686          12.2704          11.8150          11.7813   \n",
      "RNF11      ...          11.2862          11.1187          10.6101          10.5163   \n",
      "\n",
      "           TCGA-MP-A4TK-01  \n",
      "sample                      \n",
      "ARHGEF10L          10.1272  \n",
      "HIF3A              10.2122  \n",
      "RNF17               0.5804  \n",
      "RNF10              11.4430  \n",
      "RNF11              11.1243  \n",
      "\n",
      "[5 rows x 576 columns]\n",
      "(20530, 576)\n"
     ]
    }
   ],
   "source": [
    "expression = pd.read_csv(Path(data_dir_raw) / 'expression.tsv', sep='\\t', index_col=0)\n",
    "print(expression.head())\n",
    "print(expression.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac5417ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          sampleID  ABSOLUTE_Ploidy  ABSOLUTE_Purity  AKT1 ALK_translocation  ...  \\\n",
      "0  TCGA-05-4244-01              NaN              NaN   NaN               NaN  ...   \n",
      "1  TCGA-05-4249-01             3.77             0.46  none               NaN  ...   \n",
      "2  TCGA-05-4250-01              NaN              NaN   NaN               NaN  ...   \n",
      "3  TCGA-05-4382-01              NaN              NaN  none               NaN  ...   \n",
      "4  TCGA-05-4384-01             2.04             0.48  none               NaN  ...   \n",
      "\n",
      "             _GENOMIC_ID_TCGA_LUAD_RPPA _GENOMIC_ID_TCGA_LUAD_exp_HiSeqV2_percentile  \\\n",
      "0  59881fe3-70f4-4fa0-9355-f50250872752         e6a101b9-61f9-4ed1-a59f-d9db3fdb4555   \n",
      "1  5b46a235-49a6-43ea-91b6-82a384702477         d1a8d88d-1708-4959-9695-6f2e67853bd5   \n",
      "2  3f069d62-c371-4f30-bdb4-c58093bbb1d7         bba9333a-09f7-4585-b22e-e4ae4049f7da   \n",
      "3                                   NaN         e4177b01-6898-4bb7-b38d-0c09f85c5668   \n",
      "4  5e735c28-5f2b-47ec-98fb-2f57cc8a5d6a         7d6cf896-b04a-431a-a192-aaf540eeaf77   \n",
      "\n",
      "  _GENOMIC_ID_TCGA_LUAD_mutation_broad_gene _GENOMIC_ID_data/public/TCGA/LUAD/miRNA_HiSeq_gene  \\\n",
      "0              TCGA-05-4244-01A-01D-1105-08                                                NaN   \n",
      "1              TCGA-05-4249-01A-01D-1105-08                                                NaN   \n",
      "2              TCGA-05-4250-01A-01D-1105-08                                                NaN   \n",
      "3              TCGA-05-4382-01A-01D-1931-08                                                NaN   \n",
      "4              TCGA-05-4384-01A-01D-1753-08                                    TCGA-05-4384-01   \n",
      "\n",
      "  _GENOMIC_ID_TCGA_LUAD_miRNA_GA  \n",
      "0                TCGA-05-4244-01  \n",
      "1                TCGA-05-4249-01  \n",
      "2                TCGA-05-4250-01  \n",
      "3                TCGA-05-4382-01  \n",
      "4                            NaN  \n",
      "\n",
      "[5 rows x 148 columns]\n",
      "(706, 148)\n"
     ]
    }
   ],
   "source": [
    "clinical = pd.read_csv(Path(data_dir_raw) / 'clinical.tsv', sep='\\t', index_col=0)\n",
    "print(clinical.head())\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0047757",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            sample      _PATIENT  OS  OS.time  DSS  ...  DFI  DFI.time  PFI  PFI.time  Redaction\n",
      "0  TCGA-05-4244-01  TCGA-05-4244   0      0.0  0.0  ...  NaN       NaN    0       0.0        NaN\n",
      "1  TCGA-05-4249-01  TCGA-05-4249   0   1523.0  0.0  ...  NaN       NaN    0    1523.0        NaN\n",
      "2  TCGA-05-4250-01  TCGA-05-4250   1    121.0  NaN  ...  NaN       NaN    0     121.0        NaN\n",
      "3  TCGA-05-4382-01  TCGA-05-4382   0    607.0  0.0  ...  1.0     334.0    1     334.0        NaN\n",
      "4  TCGA-05-4384-01  TCGA-05-4384   0    426.0  0.0  ...  NaN       NaN    1     183.0        NaN\n",
      "\n",
      "[5 rows x 11 columns]\n",
      "(641, 11)\n"
     ]
    }
   ],
   "source": [
    "survival = pd.read_csv(Path(data_dir_raw) / 'survival.tsv', sep='\\t', index_col=0)\n",
    "print(survival.head())\n",
    "print(survival.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416e110e",
   "metadata": {},
   "source": [
    "# reshape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9869da84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(576, 20530)\n"
     ]
    }
   ],
   "source": [
    "# transpose expression so index = samples and columns = genes\n",
    "expression = expression.T  \n",
    "expression.index.name = None\n",
    "expression.columns.name = None\n",
    "print(expression.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "516a303d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(706, 147)\n"
     ]
    }
   ],
   "source": [
    "#Â set clinical index = sampleID\n",
    "clinical = clinical.set_index('sampleID') \n",
    "clinical.index.name = None\n",
    "clinical.columns.name = None\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "63d6bd6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(641, 10)\n"
     ]
    }
   ],
   "source": [
    "# set survival index = sample\n",
    "survival = survival.set_index('sample') \n",
    "survival.index.name = None\n",
    "survival.columns.name = None\n",
    "print(survival.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa9fa8e",
   "metadata": {},
   "source": [
    "# drop rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "87ffc23b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(632, 10)\n"
     ]
    }
   ],
   "source": [
    "# remove rows in survival where OS (status_col) or OS.time (time_col) is missing\n",
    "survival = survival.dropna(subset=['OS', 'OS.time'])\n",
    "print(survival.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "43dc2161",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(642, 147)\n"
     ]
    }
   ],
   "source": [
    "# only use informed_consent_verified=YES\n",
    "clinical = clinical.loc[clinical['informed_consent_verified'] == 'YES', :]\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d91b8f7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_type_id  sample_type        \n",
      "1.0             FFPE Scrolls             1\n",
      "                Primary Tumor          519\n",
      "2.0             Recurrent Tumor          2\n",
      "11.0            Solid Tissue Normal    120\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# check sample id codes in clinical\n",
    "print(clinical.value_counts(['sample_type_id', 'sample_type']).sort_index())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96da2e12",
   "metadata": {},
   "source": [
    "See sample id code definitions:\n",
    "\n",
    "https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aa9a7aee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(522, 147)\n"
     ]
    }
   ],
   "source": [
    "# drop normal samples\n",
    "sample_type_mask = ~(clinical['sample_type_id'] == 11.0)\n",
    "clinical = clinical.loc[sample_type_mask, :]\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8461112a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_type_id</th>\n",
       "      <th>sample_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-50-5066-01</th>\n",
       "      <td>5066</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-50-5066-02</th>\n",
       "      <td>5066</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-50-5946-01</th>\n",
       "      <td>5946</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-50-5946-02</th>\n",
       "      <td>5946</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                patient_id  sample_type_id      sample_type\n",
       "TCGA-50-5066-01       5066             1.0    Primary Tumor\n",
       "TCGA-50-5066-02       5066             2.0  Recurrent Tumor\n",
       "TCGA-50-5946-01       5946             1.0    Primary Tumor\n",
       "TCGA-50-5946-02       5946             2.0  Recurrent Tumor"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check only one sample per patient\n",
    "mask = clinical.duplicated('patient_id', keep=False)\n",
    "clinical\\\n",
    "    .loc[mask, ['patient_id', 'sample_type_id', 'sample_type']]\\\n",
    "    .sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bd22b013",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_type_id\n",
      "1.0    520\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# keep the primary tumour samples for these patients\n",
    "clinical = clinical.drop_duplicates('patient_id', keep='first')\n",
    "assert clinical['patient_id'].nunique() == clinical.shape[0]\n",
    "print(clinical.value_counts('sample_type_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b9e0ac11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(506, 20530)\n",
      "(506, 147)\n",
      "(506, 10)\n"
     ]
    }
   ],
   "source": [
    "# use samples common to all data sets\n",
    "samples = expression.index\\\n",
    "    .intersection(survival.index)\\\n",
    "    .intersection(clinical.index)\n",
    "\n",
    "expression = expression.loc[samples, :]\n",
    "clinical = clinical.loc[samples, :]\n",
    "survival = survival.loc[samples, :]\n",
    "\n",
    "assert (expression.index == clinical.index).all()\n",
    "assert (expression.index == survival.index).all()\n",
    "\n",
    "print(expression.shape)\n",
    "print(clinical.shape)\n",
    "print(survival.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602d254e",
   "metadata": {},
   "source": [
    "# drop columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "abaeb734",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(506, 101)\n"
     ]
    }
   ],
   "source": [
    "# drop redundant columns\n",
    "cols_to_drop = [\n",
    "    # used to derive event col (OS) and time col (OS.time) \n",
    "    'vital_status', \n",
    "    'days_to_death',\n",
    "    'days_to_last_followup', \n",
    "    # duplicate of age_at_initial_pathologic_diagnosis\n",
    "    'days_to_birth', \n",
    "    'years_to_birth', \n",
    "    # not relevant to this analysis\n",
    "    'new_tumor_event_after_initial_treatment', # not modelling recurrence\n",
    "    'followup_case_report_form_submission_reason', # not modelling recurrence\n",
    "    'is_ffpe', # not analysing molecular data sensitive to FFPE samples\n",
    "    'tissue_prospective_collection_indicator',\n",
    "    'tissue_retrospective_collection_indicator',\n",
    "    # already used for filtering and of no further use\n",
    "    'informed_consent_verified', \n",
    "    'sample_type_id',\n",
    "    'sample_type',\n",
    "    # administrative etc\n",
    "    'bcr_followup_barcode',\n",
    "    'bcr_patient_barcode',\n",
    "    'bcr_sample_barcode',\n",
    "    'form_completion_date',\n",
    "    'pathology_report_file_name',\n",
    "    'patient_id',\n",
    "    'system_version',\n",
    "    'tissue_source_site',\n",
    "    'vial_number',\n",
    "    'WGS_as_of_20120731_0_no_1_yes',\n",
    "    '_INTEGRATION',\n",
    "    '_PATIENT',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_mutation',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_mutation_curated_broad_gene',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_PDMarray',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_exp_HiSeqV2',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_G4502A_07_3',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_hMethyl27',\n",
    "    '_GENOMIC_ID_data/public/TCGA/LUAD/miRNA_GA_gene',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_gistic2',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_hMethyl450',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_PDMRNAseqCNV',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_gistic2thd',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_PDMarrayCNV',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_exp_HiSeqV2_exon',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_miRNA_HiSeq',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_RPPA_RBN',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_exp_HiSeqV2_PANCAN',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_PDMRNAseq',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_RPPA',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_exp_HiSeqV2_percentile',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_mutation_broad_gene',\n",
    "    '_GENOMIC_ID_data/public/TCGA/LUAD/miRNA_HiSeq_gene',\n",
    "    '_GENOMIC_ID_TCGA_LUAD_miRNA_GA',\n",
    "]\n",
    "\n",
    "column_mask = ~clinical.columns.isin(cols_to_drop)\n",
    "clinical = clinical.loc[:, column_mask]\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c1b91cbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(506, 10)\n"
     ]
    }
   ],
   "source": [
    "# drop columns from survival that are already in expression or clinincal\n",
    "column_mask = ~survival.columns.isin(expression.columns.union(clinical.columns))\n",
    "survival = survival.loc[:, column_mask]\n",
    "print(survival.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5e293129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(506, 86)\n"
     ]
    }
   ],
   "source": [
    "# drop columns from clinical that are already in expression\n",
    "column_mask = ~clinical.columns.isin(expression.columns)\n",
    "clinical = clinical.loc[:, column_mask]\n",
    "print(clinical.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d24c653",
   "metadata": {},
   "source": [
    "# save processed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53c51ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression.to_csv(Path(data_dir_proc) / 'expression.tsv', sep='\\t')\n",
    "clinical.to_csv(Path(data_dir_proc) / 'clinical.tsv', sep='\\t')\n",
    "survival.to_csv(Path(data_dir_proc) / 'survival.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
